consensus/src/payload_manager.rs

```diff
@@ -1,119 +1,192 @@
+    pub async fn get_transactions(&self, block: &Block) -> Result<Vec<SignedTransaction>, Error> {
+        if block.payload().is_none() {
+            return Ok(Vec::new());
+        }
+
+        match self {
+            PayloadManager::DirectMempool => match block.payload().unwrap() {
+                Payload::InQuorumStore(_) => unreachable!("should not use QuorumStore"),
+                Payload::DirectMempool(txns) => Ok(txns.clone()),
+            },
+            PayloadManager::InQuorumStore(batch_reader, _) => {
+                match block.payload().unwrap() {
+                    Payload::DirectMempool(_) => unreachable!("should use QuorumStore"),
+                    Payload::InQuorumStore(proof_with_data) => {
+                        let status = proof_with_data.status.lock().take();
+                        match status.expect("Should have been updated before") {
+                            DataStatus::Cached(data) => {
+                                proof_with_data
+                                    .status
+                                    .lock()
+                                    .replace(DataStatus::Cached(data.clone()));
+                                Ok(data)
+                            }
+                            DataStatus::Requested(receivers) => {
+                                let mut vec_ret = Vec::new();
+                                debug!("QSE: waiting for data on {} receivers", receivers.len());
+                                for (digest, rx) in receivers {
+                                    match rx.await {
+                                        Err(_) => {
+                                            // We probably advanced epoch already.
+                                            warn!("Oneshot channel to get a batch was dropped");
+                                            let new_receivers = self
+                                                .request_transactions(
+                                                    proof_with_data.proofs.clone(),
+                                                    LogicalTime::new(block.epoch(), block.round()),
+                                                    batch_reader,
+                                                )
+                                                .await;
+                                            // Could not get all data so requested again
+                                            proof_with_data
+                                                .status
+                                                .lock()
+                                                .replace(DataStatus::Requested(new_receivers));
+                                            return Err(DataNotFound(digest));
+                                        }
+                                        Ok(Ok(data)) => {
+                                            debug!("QSE: got data, len {}", data.len());
```

Comment:
These `debug` messages are probably too verbose if they are printed on every request. If not ready to remove, can they be downgraded to `trace`?